### DISCUSSION TITLE
Applying transactions that modify SC state with t.Apply

### DISCUSSION THREAD

**Discussion by dankostiuk** [0xPolygon/polygon-edge#600] [General] [answered]
We are trying to implement a blockRewards system to compensate validators.

Similar to how your team added the read transaction `QueryValidators` that queries the validators() function within Staking.sol, we've tried to add an additional function that saves the blockNumber and validator address within a map defined on our custom SC - further logic within our SC then rewards validators with a set amount once per epoch.

Although our function is getting queried, the state doesn't appear to change (msg.sender isn't being stored). Is there any recommended way to do this from within polygon-edge using `t.Apply()`?

---

Earlier on we tried storing a defined blockRewardAmount within the snapshot (updated each epoch), and then tried paying out this amount within PreStateCommit() that gets called from buildBlock(), unfortunately this would work for some time until a validator would fail to verify due to an invalid block root state :

```
	// pay the block proposer the block reword from the current snapshot
	blockRewardsBonus, ok := new(big.Int).SetString(snapshot.BlockReward, 10)
	if !ok {
		return nil
	}
	txn.Txn().AddBalance(txn.GetTxContext().Coinbase, blockRewardsBonus)
```

**Comment by Kourin1996** [comment, marked as answer]
Hello, @dankostiuk.

All validator needs to do same state change. Blockchain calls `PreStateCommit` of Blockchain with the state transition object after all transactions in a block are processed. You can use this callback to update your state.

For example, staking contract is deployed without transaction at the specific height in the following code:
https://github.com/0xPolygon/polygon-edge/blob/009a813eb8242216a49d4399b6da56b1ca173ce8/consensus/ibft/pos.go#L155-L175
